### PR TITLE
support kubectl attach in KubeEdge 

### PR DESCRIPTION
**Pull request by wenchajun**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Currently, in a native kubernetes cluster, `kubectl exec` command can execute a command in a container. `kubectl logs` command can print the logs for a container in a pod. `kubectl attach` command can attach to a running container.
Users can execute these commands in the cloud and do not need to operate on the nodes.
we plan to implement `kubectl exec`, `kubectl logs` and `kubectl attach` feature based on edge computing scenarios.
These two commands(`kubectl exec`, `kubectl logs`) are currently integrated, and this PR is to integrate the `kubectl attach` command.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubeedge/kubeedge/issues/4424

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
The user can use the `kubectl attach` command to access the pod of the edge node.
```

### CODE REVIEW

**Comment by wenchajun** [comment]
@Shelley-BaoYue Can you give us some advice?

File: cloud/pkg/cloudstream/containerattach_connection.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+func (ah *ContainerAttachConnection) Serve() error {
+	defer func() {
+		close(ah.closeChan)
+		klog.V(6).Infof("%s stop successfully", ah.String())
+	}()
+
+	// first send connect message
+	connector, err := ah.SendConnection()
+	if err != nil {
+		klog.Errorf("%s send %s info error %v", ah.String(), stream.MessageTypeExecConnect, err)
```
Comment (Shelley-BaoYue) [inline cloud/pkg/cloudstream/containerattach_connection.go:106]: Try changing it to `MessageTypeAttachConnect`.

**Comment by Shelley-BaoYue** [comment]
> @Shelley-BaoYue Can you give us some advice?

I will debug it in my local environment, it may take some time to give you feedback.

File: edge/pkg/edgestream/tunnel.go [hunk 2]
```diff
@@ -104,6 +119,10 @@ func (s *TunnelSession) ServeConnection(m *stream.Message) {
 		if err := s.serveMetricsConnection(m); err != nil {
 			klog.Errorf("Serve Metrics connection error %s", m.String())
 		}
+	case stream.MessageTypeAttachConnect:
+		if err := s.serveContainerAttachConnection(m); err != nil {
+			klog.Errorf("Serve Metrics connection error %s", m.String())
```
Comment (Shelley-BaoYue) [inline edge/pkg/edgestream/tunnel.go:124]: log info error, need to be `Serve Attach`

File: pkg/stream/edgedattachconnection.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	Header   http.Header `json:"header"`
+	Method   string      `json:"method"`
+}
+
+func (ah *EdgedAttachConnection) CreateConnectMessage() (*Message, error) {
+	data, err := json.Marshal(ah)
+	if err != nil {
+		return nil, err
+	}
+	return NewMessage(ah.MessID, MessageTypeExecConnect, data), nil
```
Comment (Shelley-BaoYue) [inline pkg/stream/edgedattachconnection.go:45]: `MessageTypeExecConnect` --> `MessageTypeAttachConnect`

File: cloud/pkg/cloudstream/containerattach_connection.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+limitations under the License.
+*/
+
+package cloudstream
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/kubeedge/kubeedge/common/constants"
```
Comment (Shelley-BaoYue) [inline cloud/pkg/cloudstream/containerattach_connection.go:23]: The import sequence needs to be adjusted. Generally, the import sequence is the standard lib, third-party lib, and local 
 lib.

File: pkg/stream/edgedattachconnection.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+*/
+
+package stream
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
```
Comment (Shelley-BaoYue) [inline pkg/stream/edgedattachconnection.go:24]: Adjusting the import sequence

File: pkg/stream/edgedattachconnection.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		select {
+		case <-ah.Stop:
+		default:
+			return
+		}
+	}
+}
+
+func (ah *EdgedAttachConnection) receiveFromCloudStream(con net.Conn, stop chan struct{}) {
+	// todo: read msg from cloudstream
```
Comment (Shelley-BaoYue) [inline pkg/stream/edgedattachconnection.go:75]: `todo` flag can be removed now

File: pkg/stream/edgedattachconnection.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+			if err != nil {
+				klog.Errorf("failed to write, err: %v", err)
+			}
+		}
+	}
+	klog.V(6).Infof("%s read channel closed", ah.String())
+}
+
+func (ah *EdgedAttachConnection) write2CloudStream(tunnel SafeWriteTunneler, con net.Conn, stop chan struct{}) {
+	// todo: write response to tunnel
```
Comment (Shelley-BaoYue) [inline pkg/stream/edgedattachconnection.go:93]: ditto

File: pkg/stream/edgedattachconnection.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+			klog.Errorf("%v failed to write to tunnel, msg: %+v, err: %v", ah.String(), msg, err)
+			return
+		}
+		klog.V(6).Infof("%v write attach data %v", ah.String(), data[:n])
+	}
+
+}
+
+func (ah *EdgedAttachConnection) Serve(tunnel SafeWriteTunneler) error {
+	// todo: serve msg from clousstream, send req to edged and write response to tunnel
```
Comment (Shelley-BaoYue) [inline pkg/stream/edgedattachconnection.go:118]: ditto

**Comment by Shelley-BaoYue** [comment]
@wenchajun  I have successfully debugged locally. You can adjust it according to my modification suggestions and try again.

**Comment by wenchajun** [comment]
> @wenchajun I have successfully debugged locally. You can adjust it according to my modification suggestions and try again.

Thanks for your reply, I have fixed the problems.

**Comment by wenchajun** [comment]
In the CI just now, I see that the error is reported as an extra blank line added. I modified it, can you help me rerun this CI? @Shelley-BaoYue

**Comment by Shelley-BaoYue** [comment]
> In the CI just now, I see that the error is reported as an extra blank line added. I modified it, can you help me rerun this CI? @Shelley-BaoYue

OK. And the description of this pr could be added,  including associated issues #4424 , proposal: #4690 .

File: pkg/stream/constants.go [hunk 9]
```diff
@@ -8,6 +8,7 @@ const (
 const (
 	MessageTypeLogsConnect MessageType = iota
 	MessageTypeExecConnect
+	MessageTypeAttachConnect
```
Comment (Shelley-BaoYue) [inline pkg/stream/constants.go:11]: Here is a compatibility problem. We hope CloudCore can be compatible with EdgeCore of earlier versions generally.  If `MessageTypeAttachConnect` is inserted here, when CloudCore send `MessageTypeData`，EdgeCore will identify it as `MessageTypeRemoveConnect`. I suggest adding `MessageTypeAttachConnect` at the end, and the judgment at   line 195 of `edge/pkg/edgestream/tunnel` needs to be changed accordingly.

Comment (wenchajun) [reply pkg/stream/constants.go]: fixed

File: edge/pkg/edgestream/tunnel.go [hunk 10]
```diff
@@ -172,7 +191,7 @@ func (s *TunnelSession) Serve() error {
 			return fmt.Errorf("close tunnel stream connection, error:%s", string(mess.Data))
 		}
 
-		if mess.MessageType < stream.MessageTypeData {
+		if (mess.MessageType < stream.MessageTypeData) || (mess.MessageType == stream.MessageTypeAttachConnect) {
```
Comment (Shelley-BaoYue) [inline edge/pkg/edgestream/tunnel.go:194]: `mess.MessageType >= stream.MessageTypeAttachConnect` maybe better for expansion

Comment (Shelley-BaoYue) [reply edge/pkg/edgestream/tunnel.go:194]: > ```
> const (
> 	MessageTypeLogsConnect MessageType = iota
> 	MessageTypeExecConnect
> 	MessageTypeMetricConnect
> 	MessageTypeData
> 	MessageTypeRemoveConnect
> 	MessageTypeCloseConnect
> 	MessageTypeAttachConnect
> ```
> 
> What should this field be adjusted to be greater than or equal to if it is increased by 1 each time here?

I guess that in order to ensure compatibility, new message type will be added at the end. If we add more capabilities of kubectl in the future, we don't need to change that judgment, just using `mess.MessageType >= stream.MessageTypeAttachConnect` is OK.  It's just my suggestion.

Comment (wenchajun) [reply edge/pkg/edgestream/tunnel.go]: I get it .thanks

**Comment by wenchajun** [comment]
All changes have been made, are there any questions so far? @Shelley-BaoYue

**Comment by wenchajun** [comment]
It looks like there is a slight error in action, please check https://github.com/codecov/codecov-action/issues/598

**Comment by Shelley-BaoYue** [comment]
PTAL  @fisherxu @wackxu

**Pull request merged**